pkg/generator/password_reconciler.go

```diff
@@ -115,37 +125,75 @@ func (r *PasswordReconciler) createSecret(ctx context.Context, password *sgv1alp
 	return reconcile.Result{}, nil
 }
 
-type passwordParams struct {
-	Length int `yaml:"length"`
+func (r *PasswordReconciler) generate(password *sgv1alpha1.Password) (string, error) {
+	passwordVal := r.GeneratePassword(&password.Spec)
+	return passwordVal, nil
 }
 
-func (r *PasswordReconciler) generate(password *sgv1alpha1.Password) (string, error) {
-	gen := cfgtypes.NewPasswordGenerator()
+// https://golangbyexample.com/generate-random-password-golang/
+func (r *PasswordReconciler) GeneratePassword(spec *sgv1alpha1.PasswordSpec) string {
+
+	var password strings.Builder
+
+	//Set symbol character
+	for i := 0; i < spec.Symbols; i++ {
+		value, _ := randomElement(spec.SymbolCharSet)
+		password.WriteString(value)
+	}
+
+	//Set digit
+	for i := 0; i < spec.Digits; i++ {
+		value, _ := randomElement(digitSet)
+		password.WriteString(value)
+	}
 
-	genParams := passwordParams{Length: password.Spec.Length}
-	if genParams.Length == 0 {
-		genParams.Length = 40
+	//Set uppercase
+	for i := 0; i < spec.UppercaseLetters; i++ {
+		value, _ := randomElement(upperCharSet)
+		password.WriteString(value)
 	}
 
-	passwordVal, err := gen.Generate(genParams)
+	//Set lowercase
+	for i := 0; i < spec.LowercaseLetters; i++ {
+		value, _ := randomElement(lowerCharSet)
+		password.WriteString(value)
+	}
+
+	var allCharSet = lowerCharSet + upperCharSet + digitSet
+
+	remainingLength := spec.Length - spec.Symbols - spec.Digits - spec.UppercaseLetters - spec.LowercaseLetters
+	for i := 0; i < remainingLength; i++ {
+		value, _ := randomElement(allCharSet)
+		password.WriteString(value)
+	}
+
+	inRune := []rune(password.String())
+	mathrand.Shuffle(len(inRune), func(i, j int) {
```

Comment:
math/rand unfortunately is "unsuitable for security-sensitive work" per math/rand's package documentation. even though we are using crypto/rand to select individual elements, i think shuffling needs to be "secure" as well.